### PR TITLE
compiler: Move `rustc_target::spec::abi::Abi` to `rustc_abi::ExternAbi`

### PR DESCRIPTION
**Pull request by workingjubilee**
Lift `enum Abi` from its rather odd place in the middle of rustc_target, and make it available again from rustc_abi. You know, the crate where you would expect the enum that describes all the ABIs to be? The platform-neutral ones, at least. This will help further refactoring of how we handle ABIs in the near future[^0].

Rename `Abi` to `ExternAbi` because quite a lot of the compiler overloads the concept of "ABI" enough that the existing name is imprecise and it is often renamed _anyway_. Often this was to avoid conflicts with the *other* type formerly known as `Abi` (now named BackendRepr[^1]), but sometimes it is just for clarity, and this name seems more self-explanatory. It does get reexported, though, using its old name, to reduce the odds of merge-conflicting over the entire tree.

All of `ExternAbi`'s friends come along for the ride, which costs adding some optional dependencies to the rustc_abi crate. However, all of this also allows simply moving three crates entirely off rustc_target:
- rustc_hir_pretty
- rustc_lint_defs
- rustc_mir_build

This odd selection is mostly to demonstrate a secondary motivation: The majority of the front-end of the compiler should be as target-agnostic as possible, and it is easier to assure this if they simply don't depend on the crate that describes targets. Note that I didn't migrate crates that don't benefit from it in this way yet, and I didn't survey every last crate.

[^0]: This is being undertaken as part of https://github.com/rust-lang/rust/issues/119183
[^1]: https://github.com/rust-lang/rust/pull/132246

### CODE REVIEW

**Review by jieyouxu — APPROVED** [review]
Looks certainly saner to me. Will wait a bit in case anyone has some naming/terminology concerns, but otherwise LGTM.

File: compiler/rustc_abi/src/extern_abi/mod.rs [hunk 1]
```diff
@@ -7,9 +7,11 @@ use rustc_span::{Span, Symbol};
 #[cfg(test)]
 mod tests;
 
+use ExternAbi as Abi;
```
Comment (compiler-errors) [inline compiler/rustc_abi/src/extern_abi/mod.rs:10]: are we gonna get rid of this too

**Review by compiler-errors — COMMENTED** [review]
👍 👍. No naming concerns from me.

File: compiler/rustc_target/src/spec/mod.rs [hunk 2]
```diff
@@ -55,9 +55,15 @@ use crate::json::{Json, ToJson};
 use crate::spec::abi::Abi;
 use crate::spec::crt_objects::CrtObjects;
 
-pub mod abi;
 pub mod crt_objects;
 
+pub mod abi {
+    pub use rustc_abi::{
```
[the remark this answers is not in the document]
Comment (workingjubilee) [reply compiler/rustc_target/src/spec/mod.rs:61]: I can make this PR my entry in the most-files-in-tree-diffed-in-2024 competition, sure.

**Pull request merged**